File: works/tmp/17086.py
from collections import deque

N, M = map(int, input().split())
answer = 0
#arr = [[0 for _ in range(N)] for _ in range(M)]
#arr = [[] for _ in range(N)]
arr = [list(map(int, input().split())) for _ in range(N)]
mark = [[0 for _ in range(M)] for _ in range(N)]
dr = [-1, -1, 0, 1, 1, 1, 0, -1]
dc = [0, 1, 1, 1, 0, -1, -1, -1]

def bfs(r, c):
    global answer
    q = deque()
    q.append([r,c])
    mark[r][c] = 1

    while q:
        tmp = q.popleft()
        nr = tmp[0]
        nc = tmp[1]

        if arr[nr][nc] == 1:
            if mark[nr][nc]-1 > answer: answer = mark[nr][nc]-1
            break

        for i in range(8):
            nnr = nr + dr[i]
            nnc = nc + dc[i]

            if (nnr<0) | (nnr>=N) | (nnc<0) | (nnc>=M): continue
            if mark[nnr][nnc] != 0: continue
            q.append([nnr,nnc])
            mark[nnr][nnc] = mark[nr][nc] + 1
    return


for i in range(N):
    for j in range(M):
        if arr[i][j] == 1: continue
        bfs(i, j)
        mark = [[0 for _ in range(M)] for _ in range(N)]

print(answer)

File: works/tmp/17086.cpp
#include <iostream>
#include <algorithm>
#include <queue>
#include <cstring>
using namespace std;

int N, M;
int ans;
int arr[55][55];
int mark[55][55];
int dr[] = {-1, -1, 0, 1, 1, 1, 0, -1};
int dc[] = {0, 1, 1, 1, 0, -1, -1, -1};

void bfs(int r, int c)
{
    queue<pair<int,int>> q;
    q.push({r,c});
    mark[r][c] = 1;

    while(!q.empty())
    {
        int nr = q.front().first;
        int nc = q.front().second;
        q.pop();
        // 최단거리 구함
        if(arr[nr][nc] == 1)
        {
            //cout << r << ' ' << c << endl;
            //cout << nr << ' ' << nc << ", cnt = " << mark[nr][nc] << endl << endl;
            ans = max(ans, mark[nr][nc]-1);
            break;
        }
        for(int i = 0; i < 8; i++)
        {
            int nnr = nr + dr[i];
            int nnc = nc + dc[i];

            if(nnr < 0 || nnr > N || nnc < 0 || nnc > M) continue;
            if(mark[nnr][nnc] != 0) continue; // 이미 방문
            q.push({nnr,nnc});
            mark[nnr][nnc] = mark[nr][nc] + 1; // +1 칸
        }
    }
}

int main()
{
    cin >> N >> M;
    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < M; j++)
        {
            cin >> arr[i][j];
        }
    }

    for(int i = 0; i < N; i++)
    {
        for(int j = 0; j < M; j++)
        {
            if(arr[i][j] == 1) continue;
            bfs(i, j);
            fill(&mark[0][0], &mark[52][52], 0);
            //memset(mark, 0, sizeof(mark));
        }
    }

    cout << ans;
}